webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/DeploymentService.scala

```diff
@@ -104,11 +109,129 @@ trait PromiseGenerationHooks {
+   * - we want an I/O threadpool, which knows what to do when a thread is blocked for a long time,
+   *   and risks thread exaustion
+   *   (it can happens, since we have a number of files to write, some maybe big)
+   * - we want to limit the total number of concurrent execution to avoid blowming up the number
+   *   of open files, concurrent hooks, etc. This is not directly linked to the number of CPU,
+   *   but clearly there is no point having a pool of 4 threads for 1000 nodes which will
+   *   fork 1000 times for hooks - even if there is 4 threads used for that.
+   *
+   * That means that we want something looking like a sequencePar interface, but with:
+   * - a common thread pool, i/o oriented
+   * - a common max numbers of concurrent tasks.
+   *   I believe it should be common to all steps because it's mostly on the same machine - but
+   *   for now, that doesn't really matter, since each step is bloking (ie wait before next).
+   *   A parallelism around the thread pool sizing (by default, number of CPU) seems ok.
+   *
+   * here, f must not throws execption.
+   *
+   */
+  def parallelSequence[U,T](seq: Seq[U])(f:U => Box[T])(implicit parallelism: Parallelism): Box[Seq[T]] = {
+    def boxToEither(f: U => Box[T])(u:U): Either[String, T] = f(u) match {
+      case Full(t)    => Right(t)
+      case Empty      => Left("Error (no message available)")
+      case f: Failure => Left(f.messageChain)
+    }
+    def recover(u:U, ex: Throwable): Either[String, T] = Left(ex.getMessage)
+
+    parallelSequenceGen(seq)(boxToEither(f) _, recover) match {
+      case Left(nel) => Failure(nel.toList.mkString(";"))
+      case Right(x)  => Full(x)
+    }
+  }
+
+  def parallelApplicativeGen[U, T, E](seq: Seq[U])(f:U => Either[E, T], recover:(U, Throwable) => Either[E, T])(implicit parallelism: Parallelism): Seq[Either[E, T]] = {
+    import scala.concurrent.duration._
+
+    //transform the sequence in tasks, gather results unordered
+    val tasks = Task.gather(seq.map { action =>
+      parallelism.semaphore.greenLight(Task(
+        try {
+          f(action)
+        } catch {
+          case ex: Throwable => recover(action, ex)
+        }
+      ))
+    })
+
+    // give a timeout for the whole tasks sufficiently large.
+    // Hint: CF-promise taking 2s by node, for 10 000 nodes, on
+    // 4 cores => ~85 minutes...
+    // It is here mostly as a safeguard for generation which went wrong -
+    // we will already have timeout at the thread level for stalling threads.
+    val timeout = 2.hours
+
+    //exec
+    Await.result(tasks.runAsync(parallelism.scheduler), timeout)
+  }
+
+  def parallelSequenceGen[U, T, E](seq: Seq[U])(f:U => Either[E, T], recover:(U, Throwable) => Either[E, T])(implicit parallelism: Parallelism): Either[NonEmptyList[E], Seq[T]] = {
+
+    val updated = parallelApplicativeGen(seq)(f, recover)
+
+    // gather results
+    val gatherErrors = updated.foldLeft[Either[NonEmptyList[E], Seq[T]]](Right(Nil)) {
```

Comment:
Funny, errors and success will be in different ordrer (error are added as head of list, and result on tail) Don't knwo if it's a problem though ... since it's unordered and parallel i think it's ok